lib/http_get.js

```diff
@@ -0,0 +1,312 @@
+        rawKey: opContext.rawKey,
+    };
+    return { errorHandler, error: worstError };
+}
+
+/**
+ * Persists detected sub-queries errors for later clean up
+ *
+ * @param {Object} opContext - Operation context
+ * @param {Object} errorAgent - Agent to persist errors
+ * @param {Object} errorObj - Description of errors to be persisted
+ * @param {HyperdriveClient~deleteCallback} callback - Callback
+ * @param {null|Error} currentError - current final status of operation
+ * @param {boolean} callClientCb - Should we invoke callback?
+ * @return {Any} whatever callback returns
+ */
+function _handleErrors(opContext, errorAgent, errorObj,
+                       callback, currentError, callClientCb) {
+    const payload = {
+        topic: httpUtils.topics.repair,
+        key: opContext.fragments.objectKey,
+        messages: [JSON.stringify(errorObj)],
+    };
+
+    errorAgent.send([payload], err => {
+        if (err) {
+            opContext.log.error(
+                'Failure to persist fragments to repair', err);
+            // Return persistence error
+            // TODO: not sure this a proper HTTP error....
+            // TODO: must be a better way than disabling it
+            /* eslint-disable no-param-reassign */
+            err.infos = { status: 500, method: 'GET' };
+            opContext.failedToPersist = true;
+            /* eslint-enable no-param-reassign */
+            return callClientCb ? callback(err) : null;
+        }
+
+        /* Error persistence OK, return already encountered
+         * error (might be null) */
+        return callClientCb ? callback(currentError) : null;
+    });
+}
+
+function _handleCorruption(reqContext, corruptError, errorAgent) {
+    const toRepair = {
+        rawKey: reqContext.opContext.rawKey,
+        fragments: [[reqContext.chunkId, reqContext.fragmentId]],
+    };
+    _handleErrors(reqContext.opContext, errorAgent, toRepair,
+                  null, corruptError, false);
+}
+
+function getCorruptionCheckedStream(httpReply, reqContext, errorAgent) {
+    const ctypes = protocol.helpers.parseReturnedContentType(
+        httpReply.headers['content-type']);
+    const size = ctypes.get('data');
+    const expectedCRC = ctypes.get('$crc.data');
+
+    if (!size || !expectedCRC) {
+        return httpReply;
+    }
+
+    let bytesUntilCRC = size;
+    let endDataBuffer = null;
+    let crcBuffer = null;
+    const corruptSaferStream = new stream.Transform({
+        transform(chunk, encoding, callback) {
+            if (chunk.length < bytesUntilCRC) {
+                this.push(chunk);
+                bytesUntilCRC -= chunk.length;
+                callback();
+                return;
+            }
+
+            endDataBuffer = endDataBuffer === null ?
+                chunk.slice(0, bytesUntilCRC) :
+                Buffer.concat([endDataBuffer, chunk.slice(0, bytesUntilCRC)]);
+
+            crcBuffer = crcBuffer === null ?
+                chunk.slice(bytesUntilCRC) :
```

Comment:
I may be tired, but what's the point of this .slice() ? Is the CRC Buffer appended to the chunk itself ?
Is that why we also slice the chunk to get the Data part in the endDataBuffer computations ?

Reply:
Hyperdrive appends a 12 bytes payload right at the end of the data. Those 12 bytes contains 3 CRCs (we only use 1). So I need to properly split data from CRC in case the received chunk contains (part of) the CRC